sklearn/gp_search.py

```diff
@@ -0,0 +1,421 @@
+                train, test, False, params,
+                self.fit_params, return_parameters=True,
+                error_score='raise')
+                for train, test in cv]
+
+            n_test_samples = 0
+            score = 0
+            for tmp_score, tmp_n_test_samples, _, _ in cv_score:
+                tmp_score *= tmp_n_test_samples
+                n_test_samples += tmp_n_test_samples
+                score += tmp_score
+            score /= float(n_test_samples)
+
+        else:
+            score = self.estimator(params)
+
+        return score
+
+    def fit(self, X, y=None):
+        """
+        Run the hyper-parameter optimization process
+
+        Parameters
+        ----------
+        X : array-like, shape = [n_samples, n_features]
+            Training vector, where n_samples in the number of samples and
+            n_features is the number of features.
+
+        y : array-like, shape = [n_samples] or [n_samples, n_output], optional
+            Target relative to X for classification or regression;
+            None for unsupervised learning.
+
+        Returns
+        -------
+        tested_parameters_ : ndarray, the parameters tested during the process
+
+        cv_scores_ : array of the mean CV results of the parameters tested
+        """
+
+        n_tested_parameters = 0
+        tested_parameters = np.zeros((self.n_iter, self.n_parameters))
+        cv_scores = np.zeros(self.n_iter)
+        if not self._callable_estimator:
+            self.scorer_ = check_scoring(self.estimator, scoring=self.scoring)
+
+        #  Initialize with random candidates
+        init_candidates = _sample_candidates(self.n_init, self.param_bounds, self.param_is_int)
+        self.n_init = init_candidates.shape[0]
+
+        for i in range(self.n_init):
+            dict_candidate = self._vector_to_dict(init_candidates[i, :])
+            cv_score = self._evaluate_params(X, y, dict_candidate)
+
+            if self.verbose > 0:
+                print('Step ' + str(i) + ' - Hyperparameter '
+                      + str(dict_candidate) + ' ' + str(cv_score))
+
+            is_in, idx = is_in_ndarray(
+                init_candidates[i, :],
+                tested_parameters[:n_tested_parameters, :]
+            )
+
+            if not is_in:
+                tested_parameters[n_tested_parameters, :] = init_candidates[i, :]
+                cv_scores[n_tested_parameters] = cv_score
+                n_tested_parameters += 1
+            else:
+                if self.verbose > 0:
+                    print('Hyperparameter already tesed')
+                cv_scores[idx] = (cv_scores[idx] + cv_score) / 2.
+
+        for i in range(self.n_iter-self.n_init):
+
+            # Model with a Gaussian Process
+            gp = GaussianProcessRegressor(**self.gp_params)
+            gp.fit(tested_parameters[:n_tested_parameters, :],
+                   cv_scores[:n_tested_parameters])
```

Comment:
Refitting the GP in every iteration can become very slow (in particular when kernel parameters are optimized each time and n_tested_parameters becomes large). Moreover, when  n_tested_parameters is large, kernel parameters will typically not change drastically between two iterations. Thus, a more flexible schedule of GP refitting would make sense (maybe not for the first PR but later)